### PR TITLE
Mutual

### PR DESCRIPTION
**Pull request by SajjadAhmad14**
- Implement two-way friendship by adding two rows for a friendship.
- Update #create of Friendship controller.
- Update #update of Friendship controller.
- Update #destroy of Friendship controller.

### CODE REVIEW

File: app/controllers/friendships_controller.rb [hunk 1]
```diff
@@ -1,21 +1,30 @@
...
-    confirm_a_request(user)
+    friend = User.find(current_user.id)
+    @friendship = Friendship.find_by_user_id_and_friend_id(user, friend)
+    @friendship.confirmed = true
+    @friendship.save
+    @friendship = Friendship.find_by_user_id_and_friend_id(friend, user)
+    @friendship.confirmed = true
+    @friendship.save
+    redirect_to users_path
   end
```
Comment (tirthajyoti-ghosh) [inline app/controllers/friendships_controller.rb]: When a user sends an invitation (in the `#create` action), only one row is created in the friendships table with `confirmed` set to `false`. And when that invitation is accepted  (in the `#update` action), another row with inverted attributes is created with both the rows' `confirmed` attribute set to `true`.

Please rectify the code to reflect the above behavior.

File: spec/models/friendship_spec.rb [hunk 2]
```diff
@@ -2,8 +2,9 @@
 
 RSpec.describe Friendship, type: :model do
   it 'Create a new data container' do
-    friendship = Friendship.new(user_id: 1, friend_id: 2)
-    expect(friendship).to be_a_new(Friendship)
+    @friendship = Friendship.new(user_id: 1, friend_id: 2)
+    @friendship = Friendship.new(user_id: 2, friend_id: 1)
```
Comment (tirthajyoti-ghosh) [inline spec/models/friendship_spec.rb]: #### [OPTIONAL]

Declaring an instance variable here is unnecessary as you are not using them anywhere else. Please convert them to normal variables.

**Review by tirthajyoti-ghosh — CHANGES_REQUESTED** [review]
# Changes required 🛑 

## You have done a great job with this milestone 💯! Keep up the good work 👍 .

- Kudos for creating a descriptive README.

- The timeline feature is working perfectly!

### We need to work on some changes before it can be approved. Let's see...

- After I invited a user for friendship from the "all users" list, I went to that user's profile. I still see the "invite" button. This is misleading to your app's users. They might think that their invitation didn't get through. 
  - [x] Please put a "Pending invitation" or something instead of the invite button.

- I sent an invitation from "Michael" to "Ben". "Ben" should not only be able to "accept" or "reject" an invitation. But when I logged in as "Ben", I see an "invitation" button beside "accept" and "reject" buttons. 
  - [x] Please remove the "invitation" button beside a user who sent an invitation to the logged-in user.

- After I send an invitation from the "All Users" page, when I refresh the page I see the "accept" and "reject" buttons beside the user I sent an invitation to. This is again misleading the end-user. It seems like the invitation is coming from the user rather me sending an invitation to the user.
  - [x] Please remove the "accept" and "reject" buttons beside the user whom the logged-in user just sent an invitation.

- The second row with the inverted friendship will only be created when the invited user confirms the friendship, not when a user sends an invitation. I have left a comment about it below.

- The operations in `#all_friends` and `requestor` methods in "helpers/application_helper.rb" could be achieved using SQL queries or ActiveRecord methods instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby). Take a look at this 👇 

  ```ruby
  # User model

  has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
  has_many :friends, through: :confirmed_friendships
  # User model

  # Users who needs to confirm friendship
  has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", 
  foreign_key: "user_id"
  has_many :pending_friends, through: :pending_friendships, source: :friend
  # User model

  # Users who requested to be friends (needed for notifications)
  has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", 
  foreign_key: "friend_id"
  has_many :friend_requests, through: :inverted_friendships
  ```

  Since the code is not displayed in the file diff, I couldn't leave the review there.

- Add "accept" and "reject" friendship feature tests in "spec/feature/invite_spec.rb".

- Add tests for all the models (comment, post, like), not just the ones you created.

Your project looks absolutely fantastic. Just make these changes and apply for another code review.

Happy coding 👨‍💻 

*Need help? Shoot me a message on Slack - [@Ghosh](https://app.slack.com/client/T47CT8XPG/DRDM6LAQN/user_profile/URB1MHXK6)*
<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Changes Required! :recycle:
### Hi @SajjadAhmad14 :clap: :clap:
#### Good job so far :+1: However, you need to fix/implement the following requirements before this project can be approved. Check my comments below :point_down:

#### :point_right:  Missing Required Changes(from the previous review):
- _After sending an invite to a user for friendship from the "all users" list, I went to that user's profile. I still see the "invite" button. This is misleading to your app's users. They might think that their invitation didn't get through. Please put a "Pending invitation" or something that indicates a request has been sent to that either on the profile page or on the all users page. [Check it here](https://www.loom.com/share/94ac1915acbb44a4b99c06b28420d065)_
- _A scenario: "John" sent an invitation to "Dave". But when I logged in as "Dave", there is no need to add an unclickable button "NEW REQUEST"._
Screenshot from 2020-12-16 10-40-02

- _Add "accept" and "reject" friendship feature tests in "spec/feature/invite_spec.rb"._

#### _Reviewed by Temitayo Adeleke_

**Review by tirthajyoti-ghosh — APPROVED** [review]
# Approved ✅

## You have done an incredible job making this project 👏🧨💪! Just perfect 💯! 

### You have met all the requirements of the project. Go ahead and merge it! Onto the next project 🚀 !

Happy coding 👨‍💻 

<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Pull request merged**